multigeiger/ble.cpp

```diff
@@ -0,0 +1,138 @@
+// Code related to the transmission of the measured data via Bluetooth Low Energy
+// Uses GATT Heart Rate Measurement Service for notifications with CPM update
+//
+// Heart Rate Measurement = Radiation CPM, Energy Expense = Rolling Packet Counter
+//
+// Based on Neil Kolban's example file: https://github.com/nkolban/ESP32_BLE_Arduino
+// Based on Andreas Spiess' example file: https://github.com/SensorsIot/Bluetooth-BLE-on-Arduino-IDE/blob/master/Polar_H7_Sensor/Polar_H7_Sensor.ino
+
+#include <Arduino.h>
+
+#include "log.h"
+#include "userdefines.h"
+#include "utils.h"
+#include "ble.h"
+#include "display.h"
+
+#include <BLEDevice.h>
+#include <BLEServer.h>
+#include <BLEUtils.h>
+#include <BLE2902.h>
+
+#define BLE_SERVICE_HEART_RATE    BLEUUID((uint16_t)0x180D)  // 16 bit UUID of Heart Rate Service
+#define BLE_CHAR_HR_MEASUREMENT   BLEUUID((uint16_t)0x2A37)  // 16 bit UUID of Heart Rate Measurement Characteristic
+#define BLE_CHAR_HR_POSITION      BLEUUID((uint16_t)0x2A38)  // 16 bit UUID of Heart Rate Sensor Position Characteristic
+#define BLE_CHAR_HR_CONTROLPOINT  BLEUUID((uint16_t)0x2A39)  // 16 bit UUID of Heart Rate Control Point Characteristic
+#define BLE_DESCR_UUID            BLEUUID((uint16_t)0x2901)  // 16 bit UUID of BLE Descriptor
+
+BLECharacteristic bleCharHRM(BLE_CHAR_HR_MEASUREMENT, BLECharacteristic::PROPERTY_NOTIFY);
+BLEDescriptor bleDescriptorHRM(BLE_DESCR_UUID);
+BLECharacteristic bleCharHRCP(BLE_CHAR_HR_CONTROLPOINT, BLECharacteristic::PROPERTY_WRITE);
+BLEDescriptor bleDescriptorHRCP(BLE_DESCR_UUID);
+BLECharacteristic bleCharHRPOS(BLE_CHAR_HR_POSITION, BLECharacteristic::PROPERTY_READ);
+BLEDescriptor bleDescriptorHRPOS(BLE_DESCR_UUID);
+
+static bool ble_enabled = false;
+static bool device_connected = false;
+
+uint8_t flags_HRS = 0b00001001; // bit 0 --> 1 = HR (cpm) as UINT16, bit 3 --> add UINT16 Energy Expended (rolling update counter)
+uint8_t txBuffer_HRM[5]; // transmit buffer, byte[0] = flags, [1, 2] = cpm, [3, 4] = rolling update counter
+uint8_t txBuffer_HRPOS[1] = {TUBE_TYPE};
+unsigned int status_HRCP = 0;
+unsigned int cpm_update_counter = 0;
+unsigned int cpm_value = 0;
+
+bool is_ble_connected() {
+  return ble_enabled && device_connected;
```

Comment:
This way, if ble is switched off mid-operation via web interface and ble_enabled set accordingly, it will at least stop notifying and indicate correct status.

Reply:
So far, though, no call to set ble_enabled to false after switch off is implemented. Where would that call go ideally?

Reply:
not sure what the protocol is for disconnecting.

but i agree, if the web interface is used to switch off ble_enabled, we need to react somehow. that reaction might include setting `device_connected = false` though.

Reply:
The device_connected is set by the callback in the lib. We can call now setup_ble("", false) to set ble_enabled to false. But where should we set this? Do we need a check in the mainloop for the global parameter to detect its change from the web interface?

Reply:
Also, it was a bit dirty, using setup_ble(), so I've added a specific function disable_ble() to disable ble (= set ble_enable to false) after config change. Question remains - is there a callback after config was changed?